sweep/sweeper.go

```diff
@@ -454,6 +545,73 @@ func (s *UtxoSweeper) collector(blockEpochs <-chan *chainntnfs.BlockEpoch,
 	}
 }
 
+// bucketForFeeReate determines the proper bucket for a fee rate. This is done
+// in order to batch inputs with similar fee rates together.
+func (s *UtxoSweeper) bucketForFeeRate(
+	feeRate lnwallet.SatPerKWeight) (lnwallet.SatPerKWeight, error) {
+
+	step := lnwallet.SatPerKWeight(s.cfg.FeeRateBucketSize)
+	for bucket := s.relayFeeRate + step; bucket <= s.cfg.MaxFeeRate; bucket += step {
+		if feeRate <= bucket {
+			return bucket, nil
+		}
+	}
+
+	// This should never happen as we assume that the fee rate has already
+	// been validated so that it fits within our constraints, but we'll
+	// return an error anyway.
+	return 0, fmt.Errorf("invalid fee rate of %v", feeRate)
+}
+
+// clusterBySweepFeeRate takes the set of pending inputs within the UtxoSweeper
+// and clusters those together with similar fee rates. Each cluster contains a
+// sweep fee rate, which is determined by calculating the average fee rate of
+// all inputs within that cluster.
+func (s *UtxoSweeper) clusterBySweepFeeRate() []inputCluster {
+	bucketInputs := make(map[lnwallet.SatPerKWeight]pendingInputs)
+	inputFeeRates := make(map[wire.OutPoint]lnwallet.SatPerKWeight)
+
+	// First, we'll group together all inputs with similar fee rates. This
+	// is done by determining the fee rate bucket they should belong in.
+	for op, input := range s.pendingInputs {
+		feeRate, err := s.validateFeePreference(input.feePreference)
+		if err != nil {
+			log.Warnf("Skipping input %v: %v", op, err)
+			continue
+		}
```

Comment:
I understood this should never be possible? The static fee rate has been validated already when the input was submitted to the sweeper and the fee estimate should never go out of the ranges?

Reply:
Yes so we can ignore the error, but I opted to log it in case something actually goes wrong for whatever reason.